components/ledger/ledger-consensual-flow/src/main/kotlin/net/corda/ledger/consensual/flow/impl/flows/finality/ConsensualReceiveFinalityFlow.kt

```diff
@@ -46,24 +46,29 @@ class ConsensualReceiveFinalityFlow(
         val initialTransaction = session.receive<ConsensualSignedTransactionInternal>()
         val transactionId = initialTransaction.id
 
+        verifyExistingSignatures(initialTransaction)
         verifyTransaction(initialTransaction)
 
         // TODO [CORE-5982] Verify already added signatures.
-         if (validateTransaction(initialTransaction)) {
+        var transaction = if (validateTransaction(initialTransaction)) {
             log.trace { "Successfully validated transaction: $transactionId" }
             val (transaction, payload) = signTransaction(initialTransaction)
             persistenceService.persist(transaction, TransactionStatus.UNVERIFIED)
             log.debug { "Recorded transaction with the initial and our signatures: $transactionId" }
             session.send(payload)
+            transaction
         } else {
+            persistInvalidTransaction(initialTransaction)
             val payload = Payload.Failure<List<DigitalSignatureAndMetadata>>(
                 "Transaction validation failed for transaction $transactionId when signature was requested"
             )
             session.send(payload)
             throw CordaRuntimeException(payload.message)
         }
```

Comment:
With negating the if, maybe it would be a bit simpler:
```suggestion
        if (validateTransaction(initialTransaction)) {
            persistInvalidTransaction(initialTransaction)
            val payload = Payload.Failure<List<DigitalSignatureAndMetadata>>(
                "Transaction validation failed for transaction $transactionId when signature was requested"
            )
            session.send(payload)
            throw CordaRuntimeException(payload.message)
        }
        ... original if branch indented one back    
```
I may be wrong though.
I started to reorg the other bits but passing the signTransaction's return to a
`(var x, var y)` is too much fun without proper editor.

Reply:
I had a look but I don't think it really changes the code.